### ISSUE TITLE
Add -P command line option and PYTHONSAFEENV environment variable

### ISSUE THREAD

**Issue by ncoghlan** [python/cpython#57684]
BPO | [13475](https://bugs.python.org/issue13475)
--- | :---
Nosy | @gpshead, @ncoghlan, @pitrou, @kristjanvalur, @vstinner, @mattheww, @jwilk, @merwok, @ericsnowcurrently
PRs | <li>python/cpython#31542</li>
Files | <li>[issue13475_1.diff](https://bugs.python.org/file25723/issue13475_1.diff "Uploaded as text/plain at 2012-05-26.19:01:14 by @ericsnowcurrently")</li><li>[issue13475_2.diff](https://bugs.python.org/file25730/issue13475_2.diff "Uploaded as text/plain at 2012-05-26.22:20:31 by @ericsnowcurrently"): updated patch to address Antoine's comments</li><li>[issue13475_long_only.diff](https://bugs.python.org/file25759/issue13475_long_only.diff "Uploaded as text/plain at 2012-05-30.04:17:25 by @ericsnowcurrently"): switch to long options</li>

<sup>*Note: these values reflect the state of the issue at the time it was migrated and might not reflect the current state.*</sup>

<details><summary>Show more details</summary><p>

GitHub fields:
```python
assignee = None
closed_at = None
created_at = <Date 2011-11-25.00:46:33.649>
labels = ['type-feature', '3.9']
title = "Add '--mainpath'/'--nomainpath' command line options to override sys.path[0] initialisation"
updated_at = <Date 2022-02-24.02:55:12.913>
user = 'https://github.com/ncoghlan'
```

bugs.python.org fields:
```python
activity = <Date 2022-02-24.02:55:12.913>
actor = 'vstinner'
assignee = 'none'
closed = False
closed_date = None
closer = None
components = []
creation = <Date 2011-11-25.00:46:33.649>
creator = 'ncoghlan'
dependencies = []
files = ['25723', '25730', '25759']
hgrepos = []
issue_num = 13475
keywords = ['patch']
message_count = 37.0
messages = ['148295', '148305', '148310', '148336', '148350', '148378', '148379', '148394', '148433', '148458', '148552', '161676', '161677', '161685', '161686', '161689', '161691', '161695', '161696', '161749', '161926', '162677', '162708', '165164', '168428', '176160', '176522', '176526', '176527', '176529', '216789', '314041', '314193', '339624', '413868', '413871', '413877']
nosy_count = 11.0
nosy_names = ['gregory.p.smith', 'ncoghlan', 'pitrou', 'kristjan.jonsson', 'vstinner', 'mattheww', 'jwilk', 'eric.araujo', 'Arfrever', 'eric.snow', 'nisaac']
pr_nums = ['31542']
priority = 'normal'
resolution = None
stage = 'patch review'
status = 'open'
superseder = None
type = 'enhancement'
url = 'https://bugs.python.org/issue13475'
versions = ['Python 3.9']
```

</p></details>

**Comment by vstinner** [comment]
Issue discussed on python-dev:

* https://mail.python.org/archives/list/python-dev@python.org/thread/IU5Q2AXAURFVDPRWNU3BDFVKV2QX5NOR/
* https://mail.python.org/archives/list/python-dev@python.org/thread/CJ5G2HEHISXFSLSXPC4F6TNVUBE2TKIX/

**Comment by vstinner** [comment]
Discussions about this issue or about changing the default to "not add the unsafe path":

* Dec 2021: [bad magic number in six](https://github.com/benjaminp/six/issues/359#issuecomment-996159668).
  The VMware Perl SDK installation process created ``/usr/bin/six.pyc``.
* May 2020: [Python flag/envvar to not put current directory on sys.path (but don’t ignore PYTHONPATH)](https://discuss.python.org/t/python-flag-envvar-to-not-put-current-directory-on-sys-path-but-dont-ignore-pythonpath/4235)
* Mar 2018: #77234: Avoid adding an empty directory to ``sys.path`` when running a module with ``-m``.
* Jun 2017: [Python-ideas: Security: remove "." from sys.path?](https://mail.python.org/pipermail/python-ideas/2017-June/045842.html)
* Oct 2015: [CVE-2015-5652](https://www.cvedetails.com/cve/CVE-2015-5652/) about malicious ``readline.pyd`` in the current directory on Windows.
* Mar 2014: Python 3.4 adds ``-I`` option: isolated mode.
* Oct 2012: #60406: ``sys.path[0]`` security issues
* Oct 2012: Sage: [Python sys.path security risk](https://trac.sagemath.org/ticket/13579). The Sage test suite is vulnerable to stdlib module overridden in ``/tmp`` directory like ``/tmp/socket.py``.
* Nov 2011: #57715: IDLE sys.path does not contain Current Working Directory
* Nov 2011: #57684: Add ``--mainpath``/``--nomainpath`` command line options to override ``sys.path[0]`` initialisation
* Nov 2011: debian-python: [Re: ImportError: No module named multiarray (is back)](https://lists.debian.org/debian-python/2011/11/msg00058.html)
* Jun 2011: [PEP 405 -- Python Virtual Environments](https://www.python.org/dev/peps/pep-0405/). It adds the ``pyvenv.cfg`` configuration file.
* Mar 2011: [PEP 395 -- Qualified Names for Modules](https://www.python.org/dev/peps/pep-0395/)
* Aug 2009: [Python-Dev: Excluding the current path from module search path?](https://mail.python.org/pipermail/python-dev/2009-August/091360.html)
* Apr 2009: #50003: CVE-2008-5983 python: untrusted python modules search path. It adds the [PySys_SetArgvEx()](https://docs.python.org/dev/c-api/init.html#c.PySys_SetArgvEx) function to Python 2.6.6 and Python 3.1.3.
* Dec 2008: Python 3.0 is released with absolute imports by default
* May 2004: #40210: Do not add directory of sys.argv[0] into sys.path
* Dec 2003: [PEP 328 -- Imports: Multi-Line and Absolute/Relative](https://www.python.org/dev/peps/pep-0328/)

**Comment by vstinner** [comment]
IPython uses a similar workaround:

```py
    # Remove the CWD from sys.path while we load stuff.
    # This is added back by InteractiveShellApp.init_path()
    if sys.path[0] == "":
        del sys.path[0]
```

Change added to fix a common mistake: an user got an error when running in IPython a file called ``math.py`` which overrides the stdlib ``math`` module.

* https://github.com/ipython/ipykernel/commit/3f7a03d356eee3500261acf9eec6bad2196c2097
* https://github.com/jupyter/help/issues/97
* https://github.com/ipython/ipykernel/blob/main/ipykernel_launcher.py

**Comment by vstinner** [comment]
The jsontools plugin of ohmyzsh uses ``import sys; del sys.path[0]`` without checking ``sys.path[0]`` value: [jsontools.plugin.zsh](https://github.com/ohmyzsh/ohmyzsh/blob/8f56a8bdf39d7727ab0e220f0164f78c77f9c50e/plugins/jsontools/jsontools.plugin.zsh#L48-L77).

**Comment by AA-Turner** [comment]
@vstinner is there anything left to do on this issue?

A

**Comment by vstinner** [comment]
Yes: merge or reject https://github.com/python/cpython/pull/92361 (-p option).

**Comment by kumaraditya303** [comment]
Closing as https://github.com/python/cpython/pull/92361 was rejected.

**Comment by ericsnowcurrently** [comment]
I don't think the fate of this issue is directly tied to #92361.  @vstinner, is there some reason it should be?

**Comment by merwok** [comment]
This issue was about adding `-P`, which was done, then AT asked what was left to be done and VS replied with that other PR to add `-p`, so that’s why KA closed this.

**Comment by vstinner** [comment]
Since 2011, the issue title and the proposed solution changed multiple times. I added -P command line option to Python 3.11: https://docs.python.org/dev/using/cmdline.html#cmdoption-P

It seems like adding -p is not needed only the default sys.path behavior changes. I wanted to add -p for PYTHONSAFEPATH, but I didn't convince anyone that it's required, so for now, there is no -p option.

IMO the initial issue has been fixed, so I close the issue.
